keps/sig-node/4742-node-labels-downward/README.md

```diff
@@ -0,0 +1,809 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Exposing node topology labels through the downward API provides a native method
+from which to attain node label values in pods without the need of a sidecar
+to extract information.
+
+## Motivation
+
+We’d like to change the runtime behavior of containers based on node labels.
+In our case, we’re using a CNI with DaemonSets to perform network setup, and
+would like to configure the network differently based on the presence of a node
+label (e.g. `toplogy.k8s.io/vpn=true`).
+
+A number of other use cases exist for providing node labels to pods. One
+example is utilizing topology data from cloud providers, which are automatically
+assignedto nodes, and then used to tailor workloads to minimize cross-zone data
+transfers and reduce costs. Having an easy way for pods to access these node
+topology labels would provide users a straightforward, maintainable way to
+optimize their workloads given topology constraints.
+
+Workarounds today typically involve using an initContainer to query the
+Kubernetes API and then pass data via shared volume to other containers within
+the same pod. This adds additional demand on the API server and is burdensome
+compared to the ease of using downwardAPI for pod labels and metadata.
+
+### Goals
+
+* Gain access to node labels in form of `topology.k8s.io/*` on pods through volume mounts
+* Gain access to node labels in form of `topology.k8s.io/*` on pods through environmental variables
+
+### Non-Goals
+
+* Not to expose additional node info outside of labels
+* Not to pass any additional node labels other than `topology.k8s.io/*` to pods
+* Not to guarantee the label value assigned at pod creation is the most recent node label value because it is assigned at pod creation time
+
+## Proposal
+
+The initial design includes:
+
+* Being able to pass node labels to volumes with fieldPath of `node.metadata.labels` where only labels containing `topology.k8s.io/` are passed through
```

Comment:
Is this overloading the namespace of topology.k8s.io? Are we encouraging users to overload it with their own labels (`toplogy.k8s.io/hostuefipassword=foo`)? Will users accidentally see side effects because they use the same fields inside the topology.k8s.io namespace for another purpose?

Maybe it would be better to either use a config-controlled allow-list within the kubelet(?) config or a limited allow-list of [well-known labels](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesiozone) plus a new namespace like `alpha.exposednodelabels.k8s.io/*` (feel free to come up with a better name).

Reply:
Maybe we could follow a similar logic to https://kubernetes-sigs.github.io/node-feature-discovery/v0.16/usage/features.html#built-in-labels

Reply:
If we think "topology.k8s.io" in an extensibility mechanism, we need to spec that very clearly.

e.g. `x.topology.k8s.io/<anything>` is for extensions, while `topology.k8s.io/<anything>` is reserved for k8s use.

And even then we need to define "local".  Can vendors define topology extensions?

Reply:
I'm looking to update this a KEP a bit, but i think I am a little lost.  So we need to strictly document and define through code what the `<anything>` flag can be? Is this the ask?

Reply:
flag as in part of the label that would be applied to the pod

Reply:
I'm looking for the KEP to say SOMETHING LIKE the below.  Note that this is a suggested approach, we should consider alternatives:

"""
In KEP 1659, the following labels are defined:
    * `topology.kubernetes.io/region`
    * `topology.kubernetes.io/zone`

In addition to those, KEP 1659 declares the entire `topology.kubernetes.io` prefix space as reserved for use by the Kubernetes project.

This KEP expands upon KEP1659 in the following ways:

1) The `x.topology.kubernetes.io` prefix is allocated for use by end users.  The kubernetes project itself will not define any standard labels with that prefix.
1) The `<domain>.x.topology.kubernetes.io` prefix is likewise allocated for use by end users or third-parties.  The `<domain>` portion is treated the same as a "normal" label prefix.  For example, `example.com.x.topology.k8s.io/label-name`.
3) All labels using the `topology.kubernetes.io` or `*.topology.kubernetes.io` prefix spaces are considered "safe" for workloads.  A workload may be exposed to the values of these labels which directly apply to the workload.  For example, a pod may learn the topology of the node on which it is running. 
"""

Then this KEP can describe how it will expose those labels from nodes to pods (is it a literal copy or a virtual one, who wins in case of conflict, etc).

If we think there is a need for ANOTHER prefix which is "safe" but isn't "topology", we can discuss doing a similar carve-out for another prefix.  But I'd argue it stops at 2.